File: app/src/main/java/com/craiovadata/android/sunshine/data/database/WeatherDao.java
package com.craiovadata.android.sunshine.data.database;

import java.util.Date;
import java.util.List;

import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;


/**
 * {@link Dao} which provides an api for all data operations
 */
@Dao
public interface WeatherDao {

    /**
     * +     * Selects all {@link ListWeatherEntry} entries after a give _date, inclusive. The LiveData will
     * +     * be kept in sync with the database, so that it will automatically notify observers when the
     * +     * values in the table change.
     * +     *
     * +     * @param _date A {@link Date} from which to select all future weather
     * +     * @return {@link LiveData} list of all {@link ListWeatherEntry} objects after _date
     * +
     */
    @Query("SELECT id, weatherId, date, temperature, iconCodeOWM FROM weather WHERE date >= :date " +
            "ORDER BY date ASC " +
            "LIMIT 5 ")
    LiveData<List<ListWeatherEntry>> getCurrentForecast(Date date);

    /**
     * Selects all ids entries after a give _date, inclusive. This is for easily seeing
     * what entries are in the database without pulling all of the data.
     *
     * @param date The _date to select after (inclusive)
     * @return Number of future weather forecasts stored in the database
     */
    @Query("SELECT COUNT(*) FROM weather WHERE date > :date")
    int countAllFutureWeatherEntries(Date date);

    @Query("SELECT COUNT(*) FROM weather WHERE date >= :recently AND isCurrentWeather = 1")
    int countCurrentWeather(Date recently);

    /**
     * Inserts a list of {@link WeatherEntry} into the weather table. If there is a conflicting _id
     * or _date the weather entry uses the {@link OnConflictStrategy} of replacing the weather
     * forecast. The required uniqueness of these values is defined in the {@link WeatherEntry}.
     *
     * @param weather A list of weather forecasts to insert
     */
    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void bulkInsert(WeatherEntry... weather);

    /**
     * Deletes any weather data older than the given day
     *
     * @param recently The _date to delete all prior weather from (exclusive)
     */
    @Query("DELETE FROM weather WHERE date < :recently")
    void deleteOldWeather(Date recently);

    @Query("SELECT * FROM weather WHERE date  >= :recentlyDate " +
            "ORDER BY isCurrentWeather DESC, ABS(:nowDate - date) ASC " +
            "LIMIT 1")
    LiveData<List<WeatherEntry>> getCurrentWeather_old(Date nowDate, Date recentlyDate);

    @Query("SELECT * FROM weather WHERE date  >= :recentlyDate " +
            "ORDER BY isCurrentWeather DESC, date ASC " +
            "LIMIT 1")
    LiveData<List<WeatherEntry>> getCurrentWeather(Date recentlyDate);

    @Query("SELECT * FROM weather WHERE date  >= :recentlyDate " +
            "ORDER BY isCurrentWeather DESC, ABS(:nowDate - date) ASC " +
            "LIMIT 1")
    List<WeatherEntry> getCurrentWeatherList(Date nowDate, Date recentlyDate);


    @Query("SELECT * FROM weather")
    LiveData<List<WeatherEntry>> getAllEntries();


    @Query("SELECT id, date, weatherId, iconCodeOWM, temperature FROM weather WHERE date > :tomorrowMidnightNormalizedUtc " +
            "AND (date + :offset) % (24 * :hourInMillis) " +
            "BETWEEN (11 * :hourInMillis +1) AND 14 * :hourInMillis")
    LiveData<List<ListWeatherEntry>> getMidDayForecast(Date tomorrowMidnightNormalizedUtc, long offset, long hourInMillis);


}

File: app/src/main/java/com/craiovadata/android/sunshine/data/database/WeatherDao.kt
package com.craiovadata.android.sunshine.data.database

import java.util.Date

import androidx.lifecycle.LiveData
import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.craiovadata.android.sunshine.ui.models.ListWeatherEntry
import com.craiovadata.android.sunshine.ui.models.WeatherEntry
import com.craiovadata.android.sunshine.ui.models.WebcamEntry

@Dao
interface
WeatherDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun bulkInsert(vararg weather: WeatherEntry)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    fun bulkInsertWebcams(vararg webcams: WebcamEntry)

    @Query("SELECT * FROM weather")
    fun getAllWeatherEntries(): LiveData<List<WeatherEntry>>

    @Query("SELECT * FROM weather LIMIT 1")
    fun getOneRandomWeatherEntry(): WeatherEntry?

//    @Query("SELECT id, weatherId, date, temperature, iconCodeOWM FROM weather WHERE date >= :date ORDER BY date ASC LIMIT 5")
    @Query("SELECT date, weatherId, temperature, iconCodeOWM FROM weather WHERE date >= :date ORDER BY date ASC LIMIT 5")
    fun getCurrentForecast(date: Date): LiveData<List<ListWeatherEntry>>

    @Query("SELECT COUNT(*) FROM weather WHERE date > :date")
    fun countAllFutureWeatherEntries(date: Date): Int

    @Query("SELECT COUNT(*) FROM weather WHERE date >= :recently AND isCurrentWeather = 1")
    fun countCurrentWeather(recently: Date): Int

    @Query("DELETE FROM weather WHERE date < :recently")
    fun deleteOldWeather(recently: Date)

    @Query("DELETE FROM webcams WHERE updateDate < :mDate")
    fun deleteOldWebcams(mDate: Date)

    @Query("SELECT COUNT(*) FROM webcams ")
    fun countAllWebcamEntries(): Int

    @Query("SELECT * FROM weather WHERE date  >= :recentlyDate ORDER BY isCurrentWeather DESC, date ASC LIMIT :limit")
    fun getCurrentWeather(recentlyDate: Date, limit: Int): LiveData<List<WeatherEntry>>

    @Query("SELECT * FROM weather WHERE date  >= :recentlyDate ORDER BY isCurrentWeather DESC, date ASC LIMIT 1")
    fun getCurrentWeatherList(recentlyDate: Date): List<WeatherEntry>

//    @Query("SELECT id, date, weatherId, iconCodeOWM, temperature FROM weather WHERE date > :tomorrowMidnightNormalizedUtc AND (date + :offset) % (24 * :hourInMillis) BETWEEN (11 * :hourInMillis +1) AND 14 * :hourInMillis")
    @Query("SELECT date, weatherId, iconCodeOWM, temperature FROM weather WHERE date > :tomorrowMidnightNormalizedUtc AND (date + :offset) % (24 * :hourInMillis) BETWEEN (11 * :hourInMillis +1) AND 14 * :hourInMillis")
    fun getMidDayForecast(
        tomorrowMidnightNormalizedUtc: Date,
        offset: Long,
        hourInMillis: Long
    ): LiveData<List<ListWeatherEntry>>

    @Query("SELECT * FROM webcams LIMIT 1")
    fun getLatestWebcam(): List<WebcamEntry>

    @Query("SELECT * FROM webcams")
    fun getAllWebcamEntries(): LiveData<List<WebcamEntry>>
}
